src/callbacks_step/time_series_dg2d.jl

```diff
@@ -68,6 +70,141 @@ function get_elements_by_coordinates!(element_ids, coordinates, mesh::TreeMesh,
+    element_ids .= 0
+
+    # Compute and save the barycentric coordinate on each element
+    bary_centers = zeros(eltype(mesh.corners), 2, mesh.n_elements)
+    calc_bary_centers!(bary_centers, dg, cache)
+
+    # Iterate over coordinates
+    distances = zeros(eltype(mesh.corners), mesh.n_elements)
+    indices = zeros(Int, mesh.n_elements, 2)
+    for index in 1:length(element_ids)
+        # Grab the current point for which the element needs found
+        point = SVector(coordinates[1, index],
+                        coordinates[2, index])
+
+        # Compute the minimum distance between the `point` and all the element surfaces
+        # saved into `distances`. The point in `node_coordinates` that gives said minimum
+        # distance on each element is saved in `indices`
+        distances, indices = calc_minimum_surface_distance(point,
+                                                           cache.elements.node_coordinates,
+                                                           dg, mesh)
+
+        # Get the candidate elements where the `point` might live
+        candidates = findall(abs.(minimum(distances) .- distances) .<
+                             500 * eps(eltype(point)))
+
+        # The minimal surface point is on a boundary so it plays no role which candidate
+        # we use to grab it. So just use the first one
+        surface_point = SVector(cache.elements.node_coordinates[1,
+                                                                indices[candidates[1],
+                                                                        1],
+                                                                indices[candidates[1],
+                                                                        2],
+                                                                candidates[1]],
+                                cache.elements.node_coordinates[2,
+                                                                indices[candidates[1],
+                                                                        1],
+                                                                indices[candidates[1],
+                                                                        2],
+                                                                candidates[1]])
+
+        # Compute the vector pointing from the current `point` toward the surface
+        P = surface_point .- point
+
+        # If the vector `P` is the zero vector then this `point` is at an element corner or
+        # on a surface. So any candidate element works, just use the first one
+        if sum(P .* P) < 500 * eps(eltype(point))
+            element_ids[index] = candidates[1]
+            continue
+        end
+
+        # Loop through all the element candidates until we find a vector from the barycenter
+        # to the surface that points in the same direction as the current `point` vector.
+        # This then gives us the correct element.
```

Comment:
Could this also be done by using the normal vectors? This could be more efficient as we wouldn't need to compute the barycenters.

Reply:
Potentially we can use the normal vectors that are already computed and stored in `cache`. There would be a bit more logic to locate on which edge from the global edge list the minimum distance point lives. I would need to test further if this is possible, but the normal vector points out of the primary into the secondary element. So one would need to be careful to make sure that dot product test is correctly done for the local coordinate system on the primary and secondary elements. Doing it this way makes it less clear how to directly extend the procedure to the `P4estMesh` because the local numbering and edge storage is slightly different. So there is a tradeoff of computing the barycenters versus introducing multiple logic checks that might slow things down.

Reply:
Okay, I was mostly just curious what would be needed to do it with normal vectors. If it is unclear whether this even leads to performance benefits and it is more complicated to adapt to other meshes, then to me it seems better to just use the current approach for now and maybe investigate this once we really need the performance.

Reply:
> Could this also be done by using the normal vectors? This could be more efficient as we wouldn't need to compute the barycenters.

This is more or less possible. I give below an alternative code for the `get_elements_by_coordinates!` function that uses the normal directions. It is a bit unwieldy as one needs to reverse engineer which edge is shared between particular candidates on an unstructured mesh. As such, this is very specific to the way the neighbour information is stored for the `UnstructuredMesh2D` type. There is one degenerate case I could not get working for this strategy. When the gauge point is very close (but not on) an element corner there might be four candidates rather than two, which breaks some of the logic in the implementation below.

<details>
<summary>
get_elements_by_coordinates!
</summary>

```julia
function get_elements_by_coordinates!(element_ids, coordinates,
                                      mesh::UnstructuredMesh2D,
                                      dg, cache)
    if length(element_ids) != size(coordinates, 2)
        throw(DimensionMismatch("storage length for element ids does not match the number of coordinates"))
    end

    # Reset element ids - 0 indicates "not (yet) found"
    element_ids .= 0

    # Iterate over coordinates
    distances = zeros(eltype(mesh.corners), mesh.n_elements)
    indices = zeros(Int, mesh.n_elements, 2)
    for index in 1:length(element_ids)
        # Grab the current point for which the element needs found
        point = SVector(coordinates[1, index],
                        coordinates[2, index])

        # Compute the minimum distance between the `point` and all the element surfaces
        # saved into `distances`. The point in `node_coordinates` that gives said minimum
        # distance on each element is saved in `indices`
        distances, indices = calc_minimum_surface_distance(point,
                                                           cache.elements.node_coordinates,
                                                           dg, mesh)

        # Get the candidate elements where the `point` might live
        candidates = findall(abs.(minimum(distances) .- distances) .<
                             500 * eps(eltype(point)))

        # The minimal surface point is on a boundary so it plays no role which candidate
        # we use to grab it. So just use the first one
        surface_point = SVector(cache.elements.node_coordinates[1,
                                                                indices[candidates[1],
                                                                        1],
                                                                indices[candidates[1],
                                                                        2],
                                                                candidates[1]],
                                cache.elements.node_coordinates[2,
                                                                indices[candidates[1],
                                                                        1],
                                                                indices[candidates[1],
                                                                        2],
                                                                candidates[1]])

        # Compute the vector pointing from the current `point` toward the surface
        P = surface_point .- point

        # If the vector `P` is the zero vector then this `point` is at an element corner or
        # on a surface. In this case the choice of a candidate element is ambiguous and
        # we just use the first candidate. However, solutions might differ at discontinuous
        # interfaces such that this choice may influence the result.
        if sum(P .* P) < 500 * eps(eltype(point))
            element_ids[index] = candidates[1]
            continue
        end

        # When there is only one candidate after the above check then the gauge point
        # is nearest to a physical boundary. So there is no neighbour and we immediately
        # take the only available candidate
        if length(candidates) == 1
            element_ids[index] = candidates[1]
            continue
        end

        # When we make it here then we have narrowed the search to two candidates.
        # Pull the second index, i.e., the edge numbers for the sides of each candidate from
        # neighbour connectivity list of the mesh
        candidate1_edges = getindex.(findall(x -> x .== candidates[1], mesh.neighbour_information[3:4, :]), [2])
        candidate2_edges = getindex.(findall(x -> x .== candidates[2], mesh.neighbour_information[3:4, :]), [2])

        # get the shared edge ID
        edge_id = intersect(candidate1_edges, candidate2_edges)

        # grab the primary element and the local side on said element from
        # the neighbour information list now that we know which shared edge
        # contains the point of minimal distance
        primary_element = getindex(mesh.neighbour_information[3, edge_id])
        secondary_element = getindex(mesh.neighbour_information[4, edge_id])
        local_edge = getindex(mesh.neighbour_information[5, edge_id])

        # grab the appropriate outward pointing normal direction on the `primary_element`
        # from the already computed values present in the `cache`. Depending on the
        # local side number we use a different index from the available indices from
        # the minimum distance computation
        if local_edge == 2 || local_edge == 4
            # fixed i so take the j index
            normal_direction = cache.elements.normal_directions[:, indices[primary_element, 2], local_edge, primary_element]
        else # local_edge == 1 || local_edge == 3
            # fixed j so take the i index
            normal_direction = cache.elements.normal_directions[:, indices[primary_element, 1], local_edge, primary_element]
        end

        # equipped with the normal direction we can now test which element "wins"
        if sum(P .* normal_direction) > zero(eltype(point))
            element_ids[index] = primary_element
        else # sum(P .* normal_direction) < zero(eltype(point))
            element_ids[index] = secondary_element
        end
    end

    return element_ids
end
```
</details>